File: rank.py
"""Deterministic property ranking against user preferences."""

from __future__ import annotations

from typing import Any


def _price(prop: dict[str, Any], intent: str) -> int | None:
    if intent == "rent":
        return prop.get("rentAED")
    return prop.get("priceAED")


def rank_properties(
    all_properties: list[dict[str, Any]],
    preferences: dict[str, Any],
) -> list[dict[str, Any]]:
    intent = preferences.get("intent", "buy")
    areas = preferences.get("areas") or []
    types = preferences.get("types") or []
    bedrooms = preferences.get("bedrooms")
    budget_max = preferences.get("budgetMax")
    styles = preferences.get("styles") or []
    must_have = [a.lower() for a in (preferences.get("mustHave") or [])]
    dealbreakers = [d.lower() for d in (preferences.get("dealbreakers") or [])]
    count = preferences.get("count") or 3

    candidates: list[tuple[int, dict[str, Any]]] = []

    for prop in all_properties:
        style = prop.get("style", "").lower()
        if style in dealbreakers:
            continue

        price = _price(prop, intent)
        if price is None:
            continue

        if areas and prop.get("area") not in areas:
            continue

        if types and prop.get("type") not in types:
            continue

        if budget_max is not None and price > budget_max * 1.05:
            continue

        score = 0

        if styles and prop.get("style") in styles:
            score += 3

        prop_beds = prop.get("bedrooms")
        if bedrooms is not None and prop_beds is not None:
            if prop_beds == bedrooms:
                score += 2
            elif abs(prop_beds - bedrooms) == 1:
                score += 1

        if types and prop.get("type") in types:
            score += 2

        prop_amenities = [a.lower() for a in prop.get("amenities", [])]
        for tag in must_have:
            if any(tag in amenity for amenity in prop_amenities):
                score += 1

        if budget_max is not None and price <= budget_max:
            ratio = price / budget_max
            score += max(0, round(2 * (1 - ratio)))

        candidates.append((score, prop))

    candidates.sort(
        key=lambda item: (
            -item[0],
            item[1].get("pricePerSqft", 0),
        )
    )

    return [prop for _, prop in candidates[:count]]


DEFAULT_PREFERENCES: dict[str, Any] = {
    "intent": "buy",
    "areas": [],
    "types": [],
    "bedrooms": None,
    "budgetMax": None,
    "styles": [],
    "mustHave": [],
    "dealbreakers": [],
    "count": 3,
    "notes": "",
}

File: lib/rank.ts
import type { Preferences, Property } from "./types";

interface ScoredProperty {
  property: Property;
  score: number;
}

export function rankProperties(properties: readonly Property[], preferences: Preferences): Property[] {
  const count = Math.max(1, preferences.count || 3);
  const dealbreakers = preferences.dealbreakers.map((item) => item.toLowerCase());
  const mustHave = preferences.mustHave.map((item) => item.toLowerCase());

  const scored: ScoredProperty[] = [];

  for (const property of properties) {
    const activePrice = preferences.intent === "rent" ? property.rentAED : property.priceAED;

    if (!activePrice) {
      continue;
    }

    if (preferences.areas.length > 0 && !preferences.areas.includes(property.area)) {
      continue;
    }

    if (preferences.budgetMax && activePrice > preferences.budgetMax * 1.05) {
      continue;
    }

    if (dealbreakers.includes(property.style.toLowerCase())) {
      continue;
    }

    const amenityText = property.amenities.join(" ").toLowerCase();
    const searchText = `${property.title} ${property.description} ${amenityText}`.toLowerCase();

    if (dealbreakers.some((dealbreaker) => dealbreaker && searchText.includes(dealbreaker))) {
      continue;
    }

    let score = 0;

    if (preferences.styles.includes(property.style)) {
      score += 3;
    }

    if (preferences.bedrooms !== undefined) {
      const bedroomDelta = Math.abs(property.bedrooms - preferences.bedrooms);
      if (bedroomDelta === 0) {
        score += 2;
      } else if (bedroomDelta === 1) {
        score += 1;
      }
    }

    if (preferences.types.includes(property.type)) {
      score += 2;
    }

    for (const requiredAmenity of mustHave) {
      if (requiredAmenity && amenityText.includes(requiredAmenity)) {
        score += 1;
      }
    }

    if (preferences.budgetMax) {
      if (activePrice <= preferences.budgetMax) {
        const closeness = activePrice / preferences.budgetMax;
        score += Math.max(0, Math.min(2, closeness * 2));
      } else {
        score += 0.25;
      }
    }

    scored.push({ property, score });
  }

  return scored
    .sort((left, right) => {
      if (right.score !== left.score) {
        return right.score - left.score;
      }

      if (left.property.pricePerSqft !== right.property.pricePerSqft) {
        return left.property.pricePerSqft - right.property.pricePerSqft;
      }

      return left.property.id.localeCompare(right.property.id);
    })
    .slice(0, count)
    .map((item) => item.property);
}
